x/observer/keeper/vote_inbound_test.go

```diff
@@ -0,0 +1,323 @@
+
+		observer := sample.AccAddress()
+		stakingMock := keepertest.GetObserverStakingMock(t, k)
+		slashingMock := keepertest.GetObserverSlashingMock(t, k)
+
+		k.SetCrosschainFlags(ctx, types.CrosschainFlags{
+			IsInboundEnabled: true,
+		})
+		k.SetChainParamsList(ctx, types.ChainParamsList{
+			ChainParams: []*types.ChainParams{
+				{
+					ChainId:     getValidEthChainIDWithIndex(t, 0),
+					IsSupported: true,
+				},
+				{
+					ChainId:     getValidEthChainIDWithIndex(t, 1),
+					IsSupported: true,
+				},
+			},
+		})
+		k.SetObserverSet(ctx, types.ObserverSet{
+			ObserverList: []string{observer},
+		})
+		stakingMock.MockGetValidator(sample.Validator(t, sample.Rand()))
+		slashingMock.MockIsTombstoned(false)
+
+		isFinalized, isNew, err := k.VoteOnInboundBallot(
+			ctx,
+			getValidEthChainIDWithIndex(t, 0),
+			getValidEthChainIDWithIndex(t, 1),
+			common.CoinType_ERC20,
+			observer,
+			"index",
+			"inTxHash",
+		)
+		require.NoError(t, err)
+
+		// ballot should be finalized since there is only one observer
+		require.True(t, isFinalized)
+		require.True(t, isNew)
+	})
+
+	t.Run("can add vote to an existing ballot", func(t *testing.T) {
+		k, ctx, _ := keepertest.ObserverKeeperWithMocks(t, keepertest.ObserverMocksAll)
+
+		observer := sample.AccAddress()
+		stakingMock := keepertest.GetObserverStakingMock(t, k)
+		slashingMock := keepertest.GetObserverSlashingMock(t, k)
+
+		k.SetCrosschainFlags(ctx, types.CrosschainFlags{
+			IsInboundEnabled: true,
+		})
+		k.SetChainParamsList(ctx, types.ChainParamsList{
+			ChainParams: []*types.ChainParams{
+				{
+					ChainId:     getValidEthChainIDWithIndex(t, 0),
+					IsSupported: true,
+				},
+				{
+					ChainId:     getValidEthChainIDWithIndex(t, 1),
+					IsSupported: true,
+				},
+			},
+		})
+		k.SetObserverSet(ctx, types.ObserverSet{
+			ObserverList: []string{observer},
+		})
+		stakingMock.MockGetValidator(sample.Validator(t, sample.Rand()))
+		slashingMock.MockIsTombstoned(false)
+
+		// set a ballot
+		threshold, err := sdk.NewDecFromStr("0.7")
+		require.NoError(t, err)
+		ballot := types.Ballot{
+			Index:            "index",
+			BallotIdentifier: "index",
+			VoterList: []string{
+				sample.AccAddress(),
+				sample.AccAddress(),
+				observer,
+				sample.AccAddress(),
+				sample.AccAddress(),
+			},
+			Votes:           types.CreateVotes(5),
+			ObservationType: types.ObservationType_InBoundTx,
+			BallotThreshold: threshold,
+			BallotStatus:    types.BallotStatus_BallotInProgress,
+		}
+		k.SetBallot(ctx, &ballot)
+
+		isFinalized, isNew, err := k.VoteOnInboundBallot(
+			ctx,
+			getValidEthChainIDWithIndex(t, 0),
+			getValidEthChainIDWithIndex(t, 1),
+			common.CoinType_ERC20,
+			observer,
+			"index",
+			"inTxHash",
+		)
+		require.NoError(t, err)
+
+		// ballot should not be finalized as the threshold is not reached
+		require.False(t, isFinalized)
+		require.False(t, isNew)
```

Comment:
Same comment as the one for outbound TX , 
Maybe we can add unit test which adds vote but doesn't finalize the ballot

Reply:
You mean the opposite? The vote is added to an existing ballot but it is not finalized?

Reply:
Looks good to me to add it. We'll need to add the extensive testing for voting logic here: https://github.com/zeta-chain/node/issues/1806

Reply:
e0395310

Reply:
Oh, I see, ballot created but not finalized? I can add this as well

Reply:
[580e897](https://github.com/zeta-chain/node/pull/1511/commits/580e897f52e5e630746686c06567784f9c00064a)